src/stackwalk.c

```diff
@@ -44,21 +60,58 @@ size_t jl_unw_stepn(bt_cursor_t *cursor, uintptr_t *ip, uintptr_t *sp, size_t ma
         ptls->safe_restore = &buf;
 #endif
         while (1) {
-           if (n >= maxsize) {
-               n = maxsize; // return maxsize + 1 if ran out of space
-               break;
-           }
-           if (!jl_unw_step(cursor, &ip[n], &thesp, &thefp))
-               break;
-           if (sp)
-               sp[n] = thesp;
-           if (add_interp_frames && jl_is_enter_interpreter_frame(ip[n])) {
-               n += jl_capture_interp_frame(&ip[n], thesp, thefp, maxsize-n-1) + 1;
-           } else {
-               n++;
-           }
+            if (n + JL_BT_MAX_ENTRY_SIZE > maxsize) {
+                // Postpone advancing the cursor: may need more space
+                need_more_space = 1;
+                break;
+            }
+            int have_more_frames = jl_unw_step(cursor, &return_ip, &thesp, &thefp);
+            if (sp)
+                sp[n] = thesp;
+            // ARM instruction pointer encoding uses the low bit as a flag for
+            // thumb mode, which must be cleared before further use. (Note not
+            // needed for ARM AArch64.) See
+            // https://github.com/libunwind/libunwind/pull/131
+            #ifdef _CPU_ARM_
+            return_ip &= ~(uintptr_t)0x1;
+            #endif
+            // For the purposes of looking up debug info for functions, we want
+            // to harvest addresses for the *call* instruction `call_ip` during
+            // stack walking.  However, this information isn't directly
+            // available. Instead, the stack walk discovers the address
+            // `return_ip` which would be *returned to* as the stack is
+            // unwound.
+            //
+            // To infer `call_ip` in full generality we would need to
+            // understand each platform ABI instruction pointer encoding and
+            // calling conventions, noting that these may vary per stack frame.
+            // (For example signal frames on linux x86_64 have `call_ip ==
+            // return_ip`.)
+            //
+            // However for our current purposes it seems sufficient to assume
+            // that `call_ip = return_ip-1`. See also:
+            //
+            // * The LLVM unwinder functions step() and setInfoBasedOnIPRegister()
+            //   https://github.com/llvm/llvm-project/blob/master/libunwind/src/UnwindCursor.hpp
+            // * The way that libunwind handles it in `unw_get_proc_name`:
+            //   https://lists.nongnu.org/archive/html/libunwind-devel/2014-06/msg00025.html
```

Comment:
As you can see, I spent a while looking up what people do for this stuff. AFAICT it seems that `return_ip - 1` should be sufficient for our use cases. libunwind does something slightly more fancy, but I guess we don't want to trace through signal handler frames. Also libunwind's `dwarf.use_prev_instr` field is hidden from us behind an opaque pointer in any case (see `unw_get_proc_name`).

The LLVM unwinder seems to have a special case for ARM in thumb mode and I tried to do the same for our ARMv7 support but I don't know how to test that yet. @yuyichao does this use of `_CPU_ARM_` look right?

Reply:
> As you can see, I spent a while looking up what people do for this stuff. AFAICT it seems that return_ip - 1 should be sufficient for our use cases. libunwind does something slightly more fancy, but I guess we don't want to trace through signal handler frames. Also libunwind's dwarf.use_prev_instr field is hidden from us behind an opaque pointer in any case (see unw_get_proc_name).

As long as we don't give that correct pointer back to libunwind and don't add the offset when we get the backtrace from the signal frame it should be fine.

> does the above use of _CPU_ARM_ look right?

LGTM.

Reply:
> don't add the offset when we get the backtrace from the signal frame

Oops, looks like I was doing that part wrong. This should now be cleaned up in the latest commit by avoiding doing the offset for contexts we know have come from a signal handler. I checked that this matches the behavior of libunwind's `unw_backtrace()` for those contexts on my machine.